fix: update spdx licenses and relax validation of license list

williamboman opened the pull request:
The license list now allows arbitrary string values, which need to be validated
through something else, for example https://www.npmjs.com/package/spdx-expression-parse.

williamboman committed:
fix: update spdx licenses and relax validation of license list

The license list now allows arbitrary string values, which need to be validated
through something else, for example https://www.npmjs.com/package/spdx-expression-parse.

chrisgrieser reviewed (approved):
Lgtm. 

But on a more general level: are licenses something we want to validate? It creates maintenance work when there are new licenses, yet it's not key to the functioning of the registry to have perfectly consistent licenses.

williamboman commented:
> But on a more general level: are licenses something we want to validate? It creates maintenance work when there are new licenses, yet it's not key to the functioning of the registry to have perfectly consistent licenses.

I'm not a lawyer but software licenses generally include stipulations surrounding redistributions of the software that requires the license to be included. In the context of a "package manager" I guess it's somewhat open to interpretation if package metadata (the registry) is subject to these license requirements.

As for licenses changing and becoming outdated in our package definitions, it's not something I've ever seen happen. For a license to change it generally needs approval of all prior authors/contributors, something that rarely is feasible in an OSS context unless CLAs have been signed by everyone. Every time I've seen an OSS project change license it has not been done correctly (meaning the old license would still apply). Again, not a lawyer but that is my assessment

Some users of Mason may also want to not use software with certain licenses, it's not too uncommon for people to exclusively use GPL software for example.

chrisgrieser commented:
I fully agree with what you say, but my point is not whether we should include licenses, but rather whether we need those licenses to be schema-checked to such a detailed degree that we add extra packages for that. 

Looking at this PR and the related other one, it feels like a lot of effort to check licenses, and also maintenance work if people come up with new licenses. 

All "just" to ensure that people enter "gpl 3.0" instead of just "gpl", a level of detail that may only really matter to 0.1% of the users, who could also just visit the website of a tool if that precise info is that relevant them.

williamboman committed:
feat: add basic pattern validation

This was my first idea but had to backtrack because capture groups
cannot be further validated. This should give a robust enough schema
that catches most incorrect licenses while also providing autocompletion
for all SPDX licenses.

williamboman commented:
Ah gotcha. Well SPDX identifiers (or rather expressions) are well established and removes any ambiguity. I can't think of any other framework to use, accepting arbitrary strings are imo a no-go ("gpl" could mean any of the available GPL licenses). As for whether they're necessary in the first place I'll refer to my previous comment.

As for validating the licenses through the new action, I don't see it being much of a maintenance effort. The JSON schema now only provides a list of SPDX identifiers for autocompletion and better diagnostic messages, they'll probably get out of date again eventually but updating them are pretty straight forward (also, virtually all packages use one of the standard licenses so inclusion of new ones would probably never become a pressing issue). The validation is now delegated to the `spdx-expression-parse` npm package which pulls valid licenses from https://github.com/jslicense/spdx-license-ids, any new license IDs adopted into SPDX will be included there.
